src/solvers/smt2_incremental/smt_to_string.cpp

```diff
@@ -229,3 +232,149 @@ std::string smt_to_string(const smt_termt &term)
   ss << term;
   return ss.str();
 }
+
+class smt_option_to_string_convertert
+  : public smt_option_const_downcast_visitort
+{
+protected:
+  std::ostream &os;
+
+public:
+  explicit smt_option_to_string_convertert(std::ostream &os) : os(os)
+  {
+  }
+
+  void visit(const smt_option_produce_modelst &produce_models) override
+  {
+    os << ":produce-models " << (produce_models.setting() ? "true" : "false");
+  }
+};
+
+std::ostream &operator<<(std::ostream &os, const smt_optiont &option)
+{
+  option.accept(smt_option_to_string_convertert{os});
+  return os;
+}
+
+std::string smt_to_string(const smt_optiont &option)
+{
+  std::stringstream ss;
+  ss << option;
+  return ss.str();
+}
+
+class smt_logic_to_string_convertert : public smt_logic_const_downcast_visitort
+{
+protected:
+  std::ostream &os;
+
+public:
+  explicit smt_logic_to_string_convertert(std::ostream &os) : os(os)
+  {
+  }
+
+  void visit(const smt_logic_quantifier_free_bit_vectorst &) override
+  {
+    os << "QF_BV";
```

Comment:
This is not the logic we currently use and this can cause errors later (we currently use `QF_AUFBV` explicitly, or for `z3` we don't specify and let `z3` assume `ALL`).

Reply:
Is `QF_BV` enough for our first slice of functionality? I plan to add more different logics in a follow up PR, but I was deliberately limiting the scope of this and the preceding PR, in order to keep things more easily reviewable.

Reply:
We need at least `UF` for free function symbols:
```
tgw@DB0907:~$ z3 -smt2 -in
(set-logic QF_BV)
(declare-fun |f| ((_ BitVec 32) (_ BitVec 33)) (_ BitVec 31))
(error "line 2 column 60: logic does not support uninterpreted functions")
```
as above `QF_BV` alone will fail, but `QF_UFBV` below is ok:
```
tgw@DB0907:~$ z3 -smt2 -in
(set-logic QF_UFBV)
(declare-fun |f| ((_ BitVec 32) (_ BitVec 33)) (_ BitVec 31))
(define-fun |f| ((|g| (_ BitVec 32)) (|h| (_ BitVec 32))) Bool (|=| |g| |h|))
(check-sat)
sat
```